hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java

```diff
@@ -612,26 +609,81 @@ private TokenPauseStatus tokenPauseStatusOf(final boolean flag) {
-			final var relationship = view.tokenAssociations().get(relKey);
-			relationships.add(new RawTokenRelationship(
-					relationship.getBalance(),
+		return getAssociatedTokenRels(view.tokenAssociations(), account, view.tokens(), maxTokensPerAccountInfo);
+	}
+
+	/**
+	 * Traverse the linkedList of the token relationships associated to the account and build a list of
+	 * TokenRelationship whose size is limited by maxTokensPerAccountInfo.
+	 *
+	 * @param allTokenAssociations
+	 * 		MerkleMap of all token associations in the state.
+	 * @param account
+	 * 		MerkelAccount for which the associations have to be looked up.
+	 * @param maxTokensPerAccountInfo
+	 * 		maximum number of token associations to be fetched.
+	 * @return
+	 * 		A list of TokenRelationships associated to the account.
+	 */
+	public static List<TokenRelationship> getAssociatedTokenRels(
+			final MerkleMap<EntityNumPair, MerkleTokenRelStatus> allTokenAssociations,
+			final MerkleAccount account,
+			final MerkleMap<EntityNum, MerkleToken> tokens,
+			final int maxTokensPerAccountInfo) {
+		final List<TokenRelationship> listOfAssociatedTokens = new ArrayList<>();
+		var currKey = account.getTokenAssociationMetadata().lastAssociation();
+		int counter = 0;
+
+		while (!currKey.equals(EntityNumPair.MISSING_NUM_PAIR) && counter < maxTokensPerAccountInfo) {
+			final var tId = currKey.asAccountTokenRel().getRight();
+			final var tokenNum = EntityNum.fromTokenId(tId);
+			final var token = tokens.getOrDefault(tokenNum, REMOVED_TOKEN);
+			final var currRel = allTokenAssociations.get(currKey);
+			listOfAssociatedTokens.add(new RawTokenRelationship(
+					currRel.getBalance(),
 					tId.getShardNum(),
 					tId.getRealmNum(),
 					tId.getTokenNum(),
-					relationship.isFrozen(),
-					relationship.isKycGranted(),
-					relationship.isAutomaticAssociation()
-			).asGrpcFor(effectiveToken));
+					currRel.isFrozen(),
+					currRel.isKycGranted(),
+					currRel.isAutomaticAssociation()
+			).asGrpcFor(token));
+			currKey = currRel.nextKey();
+			counter++;
+		}
+		return listOfAssociatedTokens;
+	}
+
+	/**
+	 * Traverse the linkedList of the token relationships associated to the account and build a list of
+	 * TokenIds whose size is limited by maxTokensPerAccountInfo.
+	 *
+	 * @param allTokenAssociations
+	 * 		MerkleMap of all token associations in the state.
+	 * @param account
+	 * 		MerkelAccount for which the associations have to be looked up.
+	 * @param maxTokensPerAccountInfo
+	 * 		maximum number of token associations to be fetched.
+	 * @return
+	 * 		A list of TokenIds associated to the account.
+	 */
+	public static List<TokenID> getAssociatedTokenIds(
+			final MerkleMap<EntityNumPair, MerkleTokenRelStatus> allTokenAssociations,
+			final MerkleAccount account,
+			final int maxTokensPerAccountInfo) {
+		final List<TokenID> listOfAssociatedTokens = new ArrayList<>();
+		var currKey = account.getTokenAssociationMetadata().lastAssociation();
```

Comment:
Small opportunity to remove duplication with `getAssociatedTokenRels()` above by adding a `MerkleTokenRelStatus` method `TokenID getLowOrderAsTokenId()` and introducing a helper like,
```
private static void doBoundedIteration(
    final MerkleMap<EntityNumPair, MerkleTokenRelStatus> allTokenAssociations,
    final EntityNumPair startKey,
    final int maxRelsToVisit,
    final BiConsumer<TokenID, MerkleTokenRelStatus> visitor
) {
    var currKey = startKey;
    var counter = 0;
    while (!currKey.equals(EntityNumPair.MISSING_NUM_PAIR) && counter < maxRelsToVisit) {
        final var rel = allTokenAssociations.get(currKey);
        final var relTokenId = currRel.getLowOrderAsTokenId(); 
	visitor.accept(relTokenId, rel);
	currKey = currRel.nextKey();
        counter++;
    }    
}
```
...then this implementation becomes,
```
    final List<TokenID> listOfAssociatedTokens = new ArrayList<>();
    final var startKey = account.getTokenAssociationMetadata().lastAssociation();

    doBoundedIteration(allTokenAssociations, startKey, maxTokensPerAccountInfo, (tokenId, rel) -> 
        listOfAssociatedTokens.add(tId));

    return listOfAssociatedTokens;
```

Reply:
+1

Reply:
Modified it to 


```
private static void doBoundedIteration(
			final MerkleMap<EntityNumPair, MerkleTokenRelStatus> allTokenAssociations,
			final MerkleMap<EntityNum, MerkleToken> tokens,
			final EntityNumPair startKey,
			final int maxRelsToVisit,
			final BiConsumer<TokenID, TokenRelationship> visitor,
			final boolean buildRelationships
	) {
		TokenRelationship relToAdd = TokenRelationship.getDefaultInstance();
		var currKey = startKey;
		var counter = 0;
		while (!currKey.equals(EntityNumPair.MISSING_NUM_PAIR) && counter < maxRelsToVisit) {
			final var currRel = allTokenAssociations.get(currKey);
			final var relTokenId = currKey.getLowOrderAsTokenId();
			if (buildRelationships && tokens != null) {
				final var tokenNum = EntityNum.fromTokenId(relTokenId);
				final var token = tokens.getOrDefault(tokenNum, REMOVED_TOKEN);
				relToAdd = new RawTokenRelationship(
						currRel.getBalance(),
						relTokenId.getShardNum(),
						relTokenId.getRealmNum(),
						relTokenId.getTokenNum(),
						currRel.isFrozen(),
						currRel.isKycGranted(),
						currRel.isAutomaticAssociation()
				).asGrpcFor(token);
			}
			visitor.accept(relTokenId, relToAdd);
			currKey = currRel.nextKey();
			counter++;
		}
	}
```

MerkleMap of `tokens` and boolean flag `buildRelationships` are toggled depending on the type of list we want.

Reply:
This works; but note you can also just have the `visitor` implementation build the `TokenRelationship` if it wants...